File: app/src/main/java/za/co/woolworths/financial/services/android/util/NotificationUtils.java
package za.co.woolworths.financial.services.android.util;

import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.content.Context;
import android.media.AudioAttributes;
import android.media.RingtoneManager;
import android.net.Uri;
import android.os.Build;
import android.util.Log;

import androidx.annotation.RequiresApi;

import com.awfs.coordination.R;
import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.GoogleApiAvailability;
import com.google.android.gms.common.api.GoogleApi;
import com.google.firebase.iid.FirebaseInstanceId;

import java.util.List;
import java.util.function.Consumer;

import retrofit2.Call;
import za.co.woolworths.financial.services.android.contracts.IResponseListener;
import za.co.woolworths.financial.services.android.models.WoolworthsApplication;
import za.co.woolworths.financial.services.android.models.dto.CreateUpdateDevice;
import za.co.woolworths.financial.services.android.models.dto.CreateUpdateDeviceResponse;
import za.co.woolworths.financial.services.android.models.network.CompletionHandler;
import za.co.woolworths.financial.services.android.models.network.OneAppService;

public class NotificationUtils {

    private static String TAG = NotificationUtils.class.getSimpleName();
    private static NotificationUtils instance;

    private final Context appContext;


    public static NotificationUtils getInstance() {
        if (instance == null)
            instance = new NotificationUtils();

        return instance;
    }

    public NotificationUtils() {
        this.appContext = WoolworthsApplication.getInstance().getApplicationContext();
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    public static void createNotificationChannelIfNeeded(Context context){
        final String channelId = context.getResources().getString(R.string.default_notification_channel_id);
        final CharSequence channelName = context.getResources().getString(R.string.default_notification_channel_name);
        final String channelDescription = context.getResources().getString(R.string.default_notification_channel_description);

        NotificationManager notificationManager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);

        if (notificationManager.getNotificationChannel(channelId) == null){
            //this notification channel was not found, remove any other channels that exists
            List<NotificationChannel> notificationChannels = notificationManager.getNotificationChannels();

            notificationChannels.forEach(notificationChannel -> notificationManager.deleteNotificationChannel(notificationChannel.getId()));
        }

        Uri soundUri = RingtoneManager.getDefaultUri(RingtoneManager.TYPE_NOTIFICATION);
        AudioAttributes audioAttributes = new AudioAttributes.Builder()
                .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
                .setUsage(AudioAttributes.USAGE_NOTIFICATION)
                .build();

        //create notification channel
        NotificationChannel notificationChannel = new NotificationChannel(channelId, channelName, NotificationManager.IMPORTANCE_HIGH);
        notificationChannel.setDescription(channelDescription);
        notificationChannel.enableVibration(true);
        notificationChannel.setVibrationPattern(new long[]{0, 500, 500});
        notificationChannel.setSound(soundUri, audioAttributes);
        notificationManager.createNotificationChannel(notificationChannel);
    }

    // Clears notification tray messages
    public static void clearNotifications(Context context) {
        NotificationManager notificationManager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        notificationManager.cancelAll();
    }

    public void sendRegistrationToServer(){
        if (isGooglePlayServicesAvailable()){
            FirebaseInstanceId.getInstance().getInstanceId().addOnCompleteListener(task -> {
                if (task.isSuccessful())
                    sendRegistrationToServer(task.getResult().getToken());
            });
        }
	}

    public void sendRegistrationToServer(String token) {

        Log.d("FCM", token);
        // sending gcm token to server
        final CreateUpdateDevice device = new CreateUpdateDevice();
        device.appInstanceId = Utils.getUniqueDeviceID(WoolworthsApplication.getInstance().getApplicationContext());
        Utils.setToken(token);
        device.pushNotificationToken = token;

        //Don't update token if pushNotificationToken or appInstanceID NULL
        if(device.appInstanceId == null || device.pushNotificationToken==null)
            return;


        //Sending Token and app instance Id to App server
        //Need to be done after Login

        Call<CreateUpdateDeviceResponse> createUpdateDeviceCall = OneAppService.INSTANCE.getResponseOnCreateUpdateDevice(device);
        createUpdateDeviceCall.enqueue(new CompletionHandler<>(new IResponseListener<CreateUpdateDeviceResponse>() {
            @Override
            public void onSuccess(CreateUpdateDeviceResponse response) {

            }

            @Override
            public void onFailure(Throwable error) {

            }
        },CreateUpdateDeviceResponse.class));
    }

    private Boolean isGooglePlayServicesAvailable() {
        // 1
        int status = GoogleApiAvailability.getInstance().isGooglePlayServicesAvailable(appContext);
        // 2

        if (status != ConnectionResult.SUCCESS) {
            // ask user to update google play services and manage the error.
            return false;
        } else {
            // 3
            return true;
        }
    }
}

File: app/src/main/java/za/co/woolworths/financial/services/android/util/NotificationUtils.kt
package za.co.woolworths.financial.services.android.util

import android.app.NotificationChannel
import android.app.NotificationManager
import android.content.Context
import android.media.AudioAttributes
import android.media.RingtoneManager
import android.os.Build
import android.util.Log
import androidx.annotation.RequiresApi
import com.awfs.coordination.R
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.GoogleApiAvailability
import com.google.android.gms.tasks.Task
import com.google.firebase.messaging.FirebaseMessaging
import dagger.hilt.android.qualifiers.ApplicationContext
import za.co.woolworths.financial.services.android.contracts.IResponseListener
import za.co.woolworths.financial.services.android.models.dao.AppInstanceObject
import za.co.woolworths.financial.services.android.models.dto.CreateUpdateDevice
import za.co.woolworths.financial.services.android.models.dto.CreateUpdateDeviceResponse
import za.co.woolworths.financial.services.android.models.network.CompletionHandler
import za.co.woolworths.financial.services.android.models.network.OneAppService.getResponseOnCreateUpdateDevice
import za.co.woolworths.financial.services.android.util.KotlinUtils.Companion.getUniqueDeviceID
import za.co.woolworths.financial.services.android.util.analytics.FirebaseManager
import javax.inject.Inject

class NotificationUtils @Inject constructor(@ApplicationContext private val context: Context) {

    @RequiresApi(api = Build.VERSION_CODES.O)
    fun createNotificationChannelIfNeeded() {
        val channelId = context.resources.getString(R.string.default_notification_channel_id)
        val channelName: CharSequence = context.resources.getString(R.string.default_notification_channel_name)
        val channelDescription = context.resources.getString(R.string.default_notification_channel_description)
        val notificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        if (notificationManager.getNotificationChannel(channelId) == null) {
            //this notification channel was not found, remove any other channels that exists
            val notificationChannels = notificationManager.notificationChannels

            notificationChannels.forEach {  notificationChannel ->
                try {
                    notificationManager.deleteNotificationChannel(notificationChannel.id)
                }catch (e : IllegalArgumentException){
                    // TODO:: Handle cannot delete default channel on android 11 and above
                    FirebaseManager.logException(e)
                }
            }
        }
        val soundUri = RingtoneManager.getDefaultUri(RingtoneManager.TYPE_NOTIFICATION)
        val audioAttributes = AudioAttributes.Builder()
            .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
            .setUsage(AudioAttributes.USAGE_NOTIFICATION)
            .build()

        //create notification channel
        val notificationChannel = NotificationChannel(channelId, channelName, NotificationManager.IMPORTANCE_HIGH)
        notificationChannel.description = channelDescription
        notificationChannel.enableVibration(true)
        notificationChannel.vibrationPattern = longArrayOf(0, 500, 500)
        notificationChannel.setSound(soundUri, audioAttributes)
        notificationManager.createNotificationChannel(notificationChannel)
    }

    // Clears notification tray messages
    fun clearNotifications() {
        val notificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        notificationManager.cancelAll()
    }

    fun sendRegistrationToServer(token: String?) {
        Log.d("FCM", token!!)
        // sending gcm token to server
        val device = CreateUpdateDevice()
        getUniqueDeviceID { deviceId: String? ->
            device.appInstanceId = deviceId
            Utils.setToken(token)
            device.pushNotificationToken = token
            device.deviceIdentityId =
                AppInstanceObject.get().currentUserObject.linkedDeviceIdentityId

            //Don't update token if pushNotificationToken or appInstanceID NULL
            if (device.appInstanceId == null || device.pushNotificationToken == null) return@getUniqueDeviceID

            //Sending Token and app instance Id to App server
            //Need to be done after Login
            val createUpdateDeviceCall =
                getResponseOnCreateUpdateDevice(device)
            createUpdateDeviceCall.enqueue(
                CompletionHandler(
                    object : IResponseListener<CreateUpdateDeviceResponse> {
                        override fun onSuccess(response: CreateUpdateDeviceResponse?) {}
                        override fun onFailure(error: Throwable?) {}
                    }, CreateUpdateDeviceResponse::class.java
                )
            )
        }
    }

    private fun isGooglePlayServicesAvailable(): Boolean {
        // 1
        val status = GoogleApiAvailability.getInstance().isGooglePlayServicesAvailable(context)
        // 2
        return status == ConnectionResult.SUCCESS
    }

    fun sendRegistrationToServer() {
        if (isGooglePlayServicesAvailable()) {
            FirebaseMessaging.getInstance().token.addOnCompleteListener { task: Task<String?> ->
                if (task.isSuccessful) sendRegistrationToServer(task.result)
            }
        }
    }
}